trieste/acquisition/multi_objective/pareto.py

```diff
@@ -73,6 +79,194 @@ def hypervolume_indicator(self, reference: TensorType) -> TensorType:
+
+        front_size = self.front.shape[0]
+
+        # Create id array to keep track of points
+        id_arr = np.expand_dims(np.arange(front_size), axis=1)
+
+        # Stitch id array, x_star and the front together
+        stitched_array = np.concatenate([id_arr, x_star, np.array(self.front)], axis=1)
+
+        # Sort array by x_star descending
+        sorted_array = stitched_array[stitched_array[:, 1].argsort()[::-1]]
+
+        samples = sorted_array[:sample_size, 2:]
+        sample_ids = sorted_array[:sample_size, 0].astype(int)
+
+        return samples, sample_ids
+
+    def _find_x_star(self, q: TensorType, p: TensorType) -> TensorType:
+
+        front_size = self.front.shape[0]
+
+        p_diag = np.expand_dims(np.diagonal(p), axis=1)
+
+        # Solve quadratic program for y*
+        P = cp.atoms.affine.wraps.psd_wrap(q)
+        G = np.eye(front_size)
+        h = np.zeros(front_size)
+        A = np.transpose(p_diag)
+        b = np.ones(1)
+
+        # Define and solve the CVXPY problem.
+        y = cp.Variable(front_size)
+        prob = cp.Problem(cp.Minimize((1 / 2) * cp.quad_form(y, P)), [G @ y >= h, A @ y == b])
+        prob.solve()
+
+        y_star = y.value
+
+        # Calculate x*
+        x_star = np.expand_dims(y_star, axis=1) / np.sum(y_star)
+
+        return x_star
+
+    def _calculate_p_matrix(self, lower_bound: TensorType, upper_bound: TensorType) -> TensorType:
+
+        front_size, front_dims = self.front.shape
+
+        p = np.zeros([front_size, front_size])
+
+        # Calculate denominator value for p matrix elements
+        denominator: float = 1
+        for i in range(front_dims):
+            if upper_bound[i] - lower_bound[i] == 0:
+                raise ValueError(
+                    "Pareto set has identical upper and lower bounds"
+                    " in a dimension, you can avoid this by setting a "
+                    "nonzero value for bounds_min_delta"
+                )
+            denominator *= upper_bound[i] - lower_bound[i]
+
+        # Fill entries of p
+        for i in range(front_size):
+            for j in range(front_size):
+                pij = 1
+                for k in range(front_dims):
+                    pij *= upper_bound[k] - max(self.front[i, k], self.front[j, k])
+                p[i, j] = pij
+
+        p = p / denominator
+
+        return p
+
+    def _get_bounds(
+        self, delta_scaling_factor: float, min_delta: float
+    ) -> tuple[TensorType, TensorType]:
+
+        front_dims = self.front.shape[1]
+
+        # Calculate the deltas to add to the bounds to get the reference point and lower bound
+        deltas = [
+            ((float(max(self.front[:, i])) - float(min(self.front[:, i]))) * delta_scaling_factor)
+            + min_delta
+            for i in range(front_dims)
+        ]
+
+        # Define lower bound and reference point
+        lower_bound = [float(min(self.front[:, i])) - deltas[i] for i in range(front_dims)]
```

Comment:
same question as above about the usage of `float()`

Reply:
Also probably unnecessary, but it means we're storing a list of floats rather than a list of numpy arrays, which is probably more natural

Reply:
maybe if this doesn't impact performance we should try to be consistent w.r.t the rest of the multiobj code in trieste?

Reply:
Do you mean convert it to tf?

Reply:
sorry that wasn't clear, maybe it makes sense to do as in `get_reference_point` for instance? (but ignore if that doesn't make sense)

Reply:
Ah yeah, it could (and should) be vectorised. I'll do that